zephyr/lib/regions_mm.c

```diff
@@ -1,54 +1,574 @@
+			heap->physical_blocks_allocators[i]->num_blocks *
+			    (heap->physical_blocks_allocators[i]->blk_sz_shift
+			     << 1))) {
+			if (heap->span) {
+				retval = sys_mem_blocks_free_contiguous(
+				    heap->physical_blocks_allocators[i], ptr,
+				    allocation->allocation_size);
+			} else {
+				retval = sys_mem_blocks_free(
+				    heap->physical_blocks_allocators[i],
+				    allocation->allocation_size, &ptr);
+			}
 			break;
-		default:
-			return -EINVAL;
 		}
 	}
 
+	list_item_del(&allocation->node);
+
+	/* Scan for changes in physical pages mapping and
+	 * apply unmap if possible
+	 */
+	struct list_item *iterator_phys;
+	struct list_item *tmp;
+	struct list_item *iterator_activ;
+	struct vmh_active_physical_mapping *physical_mAPIng;
+
+	list_for_item_safe(iterator_phys, tmp, &physical_mappings)
+	{
+		bool can_free = true;
+
+		physical_mAPIng = CONTAINER_OF(
+		    iterator_phys, struct vmh_active_physical_mapping, node);
+
+		list_for_item(iterator_activ, &heap->active_allocations) {
+			struct vmh_active_allocation *active_alloc =
+			    CONTAINER_OF(iterator_activ,
+					 struct vmh_active_allocation, node);
+
+			if (vmh_do_ranges_overlap(
+				active_alloc->allocation_ptr,
+				active_alloc->allocation_size,
+				physical_mAPIng->mapping_ptr,
+				physical_mAPIng->mapping_size)) {
+
+				can_free = false;
+				break;
+			}
+		}
+		if (can_free) {
+			sys_mm_drv_unmap_region(physical_mAPIng->mapping_ptr,
+						physical_mAPIng->mapping_size);
+			list_item_del(iterator_phys);
+			break;
+		}
+	}
+
+	rfree(physical_mAPIng);
+	rfree(allocation);
+
+	return retval;
+}
+
+/**
+ * @brief Reconfigure heap with given config
+ *
+ * This will destroy the heap from the pointer and recreate it
+ * using provided config. Individual region attribute is the
+ * "anchor" to the virtual space we use.
+ *
+ * @param heap Ptr to the heap that should be reconfigured
+ * @param cfg heap blocks configuration
+ * @param span allow contiguous on the reconfigured heap
+ *
+ * @retval heap_pointer Returns new heap pointer on success
+ * @retval NULL when reconfiguration failed
+ */
+struct virtual_memory_heap *
+vmh_reconfigure_heap(struct virtual_memory_heap *heap, struct heap_config *cfg,
+		     bool span)
+{
+	uint32_t region_attribute = heap->virtual_region->attr;
+
+	if (vmh_free_heap(heap))
+		return NULL;
+
+	return vmh_init_heap(cfg, region_attribute, span);
+}
+
+/**
+ * @brief Get default configuration for heap
+ *
+ * This will return config created based on region provided.
+ * It will try to split memory in even chunks and split it into
+ * bundles of size 64,128,...,1024 - exactly 5 sizes.
+ *
+ * @param region Virtual region which resources we will be using for heap
+ *
+ * @retval Returns default heap_config
+ */
+void vmh_get_default_heap_config(const struct sys_mm_drv_region *region,
+				 struct heap_config *cfg)
+{
+	int buffer_size = region->size / 5;
+	int block_count = buffer_size / DCACHE_LINE_SIZE;
+
```

Comment:
5 should be an macro with a significat name. Please don't use magic numbers directly in the code.